WIP: RS/AS split and PRM (#338)
modelcontextprotocol/modelcontextprotocol: docs/specification/2025-03-26/basic/authorization.mdx

implements a selected subset of their features to ensure security and interoperability
while maintaining simplicity:

- OAuth 2.1 IETF DRAFT ([draft-ietf-oauth-v2-1-12](https://datatracker.ietf.org/doc/html/draft-ietf-oauth-v2-1-12))
- OAuth 2.0 Authorization Server Metadata
  ([RFC8414](https://datatracker.ietf.org/doc/html/rfc8414))
- OAuth 2.0 Dynamic Client Registration Protocol
  ([RFC7591](https://datatracker.ietf.org/doc/html/rfc7591))
- OAuth 2.0 Protected Resource Metadata IETF DRAFT ([draft-ietf-oauth-resource-metadata-13](https://datatracker.ietf.org/doc/html/draft-ietf-oauth-resource-metadata-13))

## 2. Authorization Flow

### 2.1 Overview

1. MCP authorization servers **MUST** implement OAuth 2.1 with appropriate security
   measures for both confidential and public clients.

1. MCP authorization servers and MCP servers **SHOULD** support the OAuth 2.0 Dynamic Client Registration
   Protocol ([RFC7591](https://datatracker.ietf.org/doc/html/rfc7591)).

> 1. MCP servers and clients **MUST** implement [OAuth 2.0 Protected Resource Metadata](https://datatracker.ietf.org/doc/html/draft-ietf-oauth-resource-metadata-13)).

Reviewer: ```suggestion
MCP servers **MUST** implement
[OAuth 2.0 Protected Resource Metadata](https://datatracker.ietf.org/doc/html/draft-ietf-oauth-resource-metadata-13),
and MCP clients **MUST** be capable of consuming this metadata.
```

Author: Good call. I'll change it to 

> 1. MCP servers **MUST** implement [OAuth 2.0 Protected Resource Metadata](https://datatracker.ietf.org/doc/html/draft-ietf-oauth-resource-metadata-13).
>    MCP clients **MUST** use OAuth 2.0 Protected Resource Metadata for authorization discovery.

> 1. MCP authorization servers **SHOULD** and MCP clients **MUST** implement OAuth 2.0 Authorization

Reviewer: Is there a scenario where an AS would not support the metadata? I think we need to be opinionated and require that AS (whether third-party or standalone) support this.

```suggestion
1. MCP authorization servers **MUST** implement OAuth 2.0 Authorization
```

   Server Metadata ([RFC8414](https://datatracker.ietf.org/doc/html/rfc8414)). Servers

>    that do not support Authorization Server Metadata **MUST** follow the default URI

Reviewer: I would actually suggest that we remove this part. In either implementation, there should be AS metadata.

> 1. MCP auth implementations **MUST** implement OAuth 2.1 with appropriate security

Reviewer 2: I don't know if we want to call this an "MCP authorization server", but "OAuth authorization server supporting the MCP server" sounded a bit long to me. In any case it seemed better to name this specifically especially so that the requirement to support Authorization Server Metadata is clearer down below.

   measures for both confidential and public clients.

> 1. MCP auth implementations **SHOULD** support the OAuth 2.0 Dynamic Client Registration

Reviewer 2: ```suggestion
1. MCP authorization servers **SHOULD** support the OAuth 2.0 Dynamic Client Registration
```

   Protocol ([RFC7591](https://datatracker.ietf.org/doc/html/rfc7591)).

1. MCP servers and clients **MUST** implement [OAuth 2.0 Protected Resource Metadata](https://datatracker.ietf.org/doc/html/draft-ietf-oauth-resource-metadata-13)).

> 1. MCP servers **SHOULD** and MCP clients **MUST** implement OAuth 2.0 Authorization

Reviewer 2: ```suggestion
1. MCP authorization servers **SHOULD** and MCP clients **MUST** implement OAuth 2.0 Authorization
```

1. MCP authorization servers **MUST** implement OAuth 2.1 with appropriate security
   measures for both confidential and public clients.

> 1. MCP authorization servers and MCP servers **SHOULD** support the OAuth 2.0 Dynamic Client Registration

Reviewer: This could be re-worded:

```suggestion
MCP servers, whether implementing their own or using third-party authorization servers, **SHOULD** support OAuth 2.0 Dynamic Client Registration
```

> 1. MCP authorization servers and MCP servers **SHOULD** support the OAuth 2.0 Dynamic Client Registration

Author: Ah this should only be clients. Good catch.

   - For instance, an agent calls a secure MCP tool to check inventory at a specific
     store. No need to impersonate the end user.

### 2.2 Roles
A protected MCP server acts as a [OAuth 2.1 resource server](https://www.ietf.org/archive/id/draft-ietf-oauth-v2-1-12.html#name-roles),
capable of accepting and responding to protected resource requests using access tokens.

An MCP client acts a [OAuth 2.1 client](https://www.ietf.org/archive/id/draft-ietf-oauth-v2-1-12.html#name-roles),
making protected resource requests on behalf of a resource owner.

> The authorization server is responsible for interacting with the user and issuing access tokens for use at the MCP server. The implementation details of the authorization server are beyond the scope of this specification. It may be the same server as the

Reviewer: ```suggestion
The authorization server handles user authentication and issues access tokens that
are used when accessing MCP server resources. Specific implementation details of the
authorization server fall outside the scope of this specification.

The authorization server may be the colocated with the
```

Author: I am not sure what's the difference here. I think being clear about responsibilities is more in in with how spec languages in IETF works.

> resource server or a separate entity. Section [2.3 Authorization Server Discovery](#2-3-authorizaton-server-discovery)

Reviewer: ```suggestion
resource server or offered by a third-party provider.
Section [2.3 Authorization Server Discovery](#2-3-authorizaton-server-discovery)
```

Author: The wording is taken from the OAuth spec. I would prefer to keep it that way.

> specifies how an MCP server (OAuth 2.1 resource server) indicates the location of an OAuth 2.1 authorization server to an OAuth 2.1 client.

Reviewer: Not sure we need to be repetitive here - we've already specified that this is following the spec.

```suggestion
specifies how an MCP server indicates the location of an authorization server to a client.
```

Author: Good catch

### 2.3 Authorization Server Discovery

> This section details how MCP servers can inform MCP clients of authorization servers, and how MCP clients discover authorization servers and their capabilities.

Reviewer: ```suggestion
This section describes the mechanisms by which MCP servers advertise their associated
authorization servers to MCP clients, as well as the discovery process through which MCP
clients can determine authorization server endpoints and supported capabilities.
```

### 2.3.1 Authorization Server Location
MCP servers **MUST** implement the [OAuth 2.0 Protected Resource Metadata](https://datatracker.ietf.org/doc/html/draft-ietf-oauth-resource-metadata-13)
specification to indicate the locations of authorization servers. The Protected Resource Metadata document returned by the MCP server **MUST** include

> the `authorization_servers` field.

Reviewer: ```suggestion
the `authorization_servers` property containing at least one authorization server.
```

> The specific use of `authorization_servers` is beyond the scope of this specification; implementers should consult

Reviewer: ```suggestion
The specific use of `authorization_servers` is beyond the scope of this specification; developers should consult
```

Author: OAuth 2.1 for example mention things like:

> Implementors should carefully consult the documentation of the resource server before adopting this practice.

This feels more aligned with spec language.

the [OAuth 2.0 Protected Resource Metadata](https://datatracker.ietf.org/doc/html/draft-ietf-oauth-resource-metadata-13) documentation for

> comprehensive guidance on implementation details.

Reviewer: ```suggestion
guidance on implementation details.
```

> MCP servers **MUST** use the HTTP header `WWW-Authenticate` when returning a _401 Unauthorized_ to indicate the location of the resource server metadata URL

Reviewer: ```suggestion
MCP servers **MUST** use the `WWW-Authenticate` HTTP header when returning a `HTTP 401 Unauthorized` response to indicate the location of the Protected Resource Metadata document URL,
```

as described in [OAuth 2.0 Protected Resource Metadata](https://datatracker.ietf.org/doc/html/draft-ietf-oauth-resource-metadata-13).

> MCP clients **MUST** handle the `WWW-Authenticate` header.

Reviewer: ```suggestion
MCP clients **MUST** be able to parse `WWW-Authenticate` headers and react to
`HTTP 401 Unauthorized` responses from the MCP server.
```

#### 2.3.2 Server Metadata Discovery

MCP clients **MUST** follow the OAuth 2.0 Authorization Server Metadata protocol defined

>   in [RFC8414](https://datatracker.ietf.org/doc/html/rfc8414) to obtain the information

Reviewer: ```suggestion
in [RFC8414](https://datatracker.ietf.org/doc/html/rfc8414) to request the details
```

>   required to interact with the authorization server.

Reviewer: ```suggestion
required to interact with the authorization server.
```

#### 2.3.3 MCP specific headers for discovery

> MCP clients **SHOULD** include the header `MCP-Protocol-Version: <protocol-version>` during

Reviewer: ```suggestion
MCP clients **SHOULD** include the `MCP-Protocol-Version: <protocol-version>` HTTP header during
```

Server Metadata Discovery to allow the MCP server to respond based on the MCP protocol

> version. MCP servers **SHOULD** use the `MCP-Protocol-Version` header to determine compatibility with the MCP client.

Reviewer: ```suggestion
version.

MCP servers **SHOULD** use the `MCP-Protocol-Version` HTTP header to determine compatibility
with connected MCP clients.
```

   - For instance, an agent calls a secure MCP tool to check inventory at a specific
     store. No need to impersonate the end user.

### 2.2 Roles
A protected MCP server acts as a [OAuth 2.1 resource server](https://www.ietf.org/archive/id/draft-ietf-oauth-v2-1-12.html#name-roles),
capable of accepting and responding to protected resource requests using access tokens.

An MCP client acts a [OAuth 2.1 client](https://www.ietf.org/archive/id/draft-ietf-oauth-v2-1-12.html#name-roles),
making protected resource requests on behalf of a resource owner.

> The authorization server is beyond the scope of this specification. It may be the same server as the

Reviewer 2: ```suggestion
The authorization server is responsible for interacting with the user and issuing access tokens for use at the MCP server. The implementation details of the authorization server are beyond the scope of this specification. It may be the same server as the
```

resource server or a separate entity. Section [2.3 Authorization Server Discovery](#2-3-authorizaton-server-discovery)
specifies how an MCP server (OAuth 2.1 resource server) indicates the location of an OAuth 2.1 authorization server to an OAuth 2.1 client.

### 2.3 Authorization Server and Capability Discovery
This section details how MCP servers can inform MCP clients of authorization servers, and how MCP clients discover authorization servers and their capabilities.

### 2.3.1 Authorization Server Location
MCP servers **MUST** implement the [OAuth 2.0 Protected Resource Metadata](https://datatracker.ietf.org/doc/html/draft-ietf-oauth-resource-metadata-13)
specification to indicate the locations of authorization servers. The Protected Resource Metadata document returned by the MCP server **MUST** include
the `authorization_servers` field.

The specific use of `authorization_servers` is beyond the scope of this specification; implementers should consult
the [OAuth 2.0 Protected Resource Metadata](https://datatracker.ietf.org/doc/html/draft-ietf-oauth-resource-metadata-13) documentation for
comprehensive guidance on implementation details.

> MCP servers **MAY** use the HTTP header `WWW-Authenticate` when returning a _401 Unauthorized_ to indicate the location of the authorization

Reviewer 2: ```suggestion
MCP servers **MAY** use the HTTP header `WWW-Authenticate` when returning a _401 Unauthorized_ to indicate the location of the resource server metadata URL
```

> server as described in [OAuth 2.0 Protected Resource Metadata](https://datatracker.ietf.org/doc/html/draft-ietf-oauth-resource-metadata-13).

Reviewer 2: ```suggestion
as described in [OAuth 2.0 Protected Resource Metadata](https://datatracker.ietf.org/doc/html/draft-ietf-oauth-resource-metadata-13).
```

> MCP clients **MUST** handle the `WWW-Authenticate` header if present, and otherwise use direct discovery via the well-known URI if `WWW-Authenticate`.

Reviewer 2: ```suggestion
MCP clients **MUST** handle the `WWW-Authenticate` header if present, and otherwise use direct discovery via the well-known URI if no `WWW-Authenticate` header is present.
```

#### 2.3.2 Server Metadata Discovery

MCP clients **MUST** follow the OAuth 2.0 Authorization Server Metadata protocol defined
  in [RFC8414](https://datatracker.ietf.org/doc/html/rfc8414) to obtain the information
  required to interact with the authorization server.

#### 2.3.3 MCP specific headers for discovery

MCP clients **SHOULD** include the header `MCP-Protocol-Version: <protocol-version>` during
Server Metadata Discovery to allow the MCP server to respond based on the MCP protocol
version.

For example: `MCP-Protocol-Version: 2024-11-05`

#### 2.3.4 Sequence Diagram
The following diagram outlines an example flow:

```mermaid
sequenceDiagram
    participant C as Client
    participant M as MCP Server (Resource Server)
    participant A as Authorization Server

    C->>M: MCP Request without token
    M->>C: HTTP 401 Unauthorized with WWW-Authenticate header

>     Note over C: Extract authorization_server from WWW-Authenticate

Reviewer 2: ```suggestion
    Note over C: Extract resource_metadata from WWW-Authenticate
```

    alt WWW-Authenticate header not present
        C->>M: GET /.well-known/oauth-protected-resource

>         M->>C: Resource metadata with authorization_server URL

Reviewer 2: ```suggestion
        M->>C: Resource metadata with authorization server URL
```

The property is actually called `authorization_servers` so I took out the underscore so we can talk about just the single authorization server URL that we expect.

sequenceDiagram
    participant B as User-Agent (Browser)
    participant C as Client
    participant M as MCP Server (Resource Server)
    participant A as Authorization Server

    C->>M: MCP Request without token
    M->>C: HTTP 401 Unauthorized with WWW-Authenticate header

>     Note over C: Extract authorization_server from WWW-Authenticate

Reviewer 2: ```suggestion
    Note over C: Extract resource_metadata from WWW-Authenticate
```

If validation fails, servers **MUST** respond according to
[Section 5.3](https://datatracker.ietf.org/doc/html/draft-ietf-oauth-v2-1-12#section-5.3)
error handling requirements. Invalid or expired tokens **MUST** receive a HTTP 401

> response. MCP client **MUST NOT** send tokens other than issues by the MCP server.

Reviewer: ```suggestion
response.

MCP clients **MUST NOT** send tokens to the MCP server that were not issued by the MCP 
server's authorization server for that MCP server.
```

Author: Will use a slightly different wording.

> MCP servers **MUST** issue tokens only for valid for the MCP server. They **MUST NOT** pass

Reviewer: ```suggestion
MCP servers **MUST** only issue tokens that are valid for use with their own resources.

MCP servers **MUST NOT** accept or transit any other tokens that have an audience
```

> on any third-party token.

Reviewer: ```suggestion
different than the MCP server.
```

> 3. Token rotation **SHOULD** be implemented for enhanced security

Reviewer: Should we call out the fact that this is applicable to MCP servers that implement their own AS? Since token rotation would really be out-of-scope if we have a third-party AS.

Author: Clients should rotate as well via refresh tokens right?

> 4. Token lifetimes **SHOULD** be limited based on security requirements

Reviewer: Also applicable to when the MCP server has its own AS rather than integrate an established one.
